iree/compiler/Dialect/Flow/Transforms/PadLinalgOps.cpp

```diff
@@ -0,0 +1,156 @@
+// Copyright 2021 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "mlir/Dialect/Linalg/IR/LinalgOps.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+namespace mlir {
+namespace iree_compiler {
+namespace IREE {
+namespace Flow {
+
+namespace {
+/// A pattern to pad staticly shaped matmul operands to the next integer
+/// multiple of padSize.
+class PadMatmulOp : public OpRewritePattern<linalg::MatmulOp> {
+ public:
+  PadMatmulOp(MLIRContext *context, int size, PatternBenefit benefit = 1)
+      : OpRewritePattern<linalg::MatmulOp>(context, benefit),
+        paddingSize(size) {}
+
+  LogicalResult matchAndRewrite(linalg::MatmulOp matmulOp,
+                                PatternRewriter &rewriter) const override {
+    auto loc = matmulOp.getLoc();
+    auto lhs = matmulOp.inputs()[0];
+    auto rhs = matmulOp.inputs()[1];
+    auto result = matmulOp.outputs()[0];
+
+    if (lhs.getDefiningOp<linalg::PadTensorOp>() ||
+        rhs.getDefiningOp<linalg::PadTensorOp>()) {
+      return failure();
+    }
+
+    auto lhsType = lhs.getType().dyn_cast<RankedTensorType>();
+    auto rhsType = rhs.getType().dyn_cast<RankedTensorType>();
+
+    if (!lhsType || !rhsType) return failure();
+
+    if (!lhsType.hasStaticShape() || !rhsType.hasStaticShape()) {
+      return failure();
+    }
+
+    auto lhsShape = lhsType.getShape();
+    auto rhsShape = rhsType.getShape();
+
+    int M = lhsShape[0], K = lhsShape[1], N = rhsShape[1];
+
+    int newMSize = std::ceil(float(M) / paddingSize) * paddingSize;
+    int newNSize = std::ceil(float(N) / paddingSize) * paddingSize;
+    int newKSize = std::ceil(float(K) / paddingSize) * paddingSize;
+
+    int paddingForM = newMSize - M;
+    int paddingForN = newNSize - N;
+    int paddingForK = newKSize - K;
+
+    if (paddingForM == 0 && paddingForN == 0 && paddingForK == 0)
+      return failure();
+    auto elementType = lhsType.getElementType();
+
+    auto lhsPaddedType =
+        RankedTensorType::get({newMSize, newKSize}, elementType);
+
+    auto rhsPaddedType =
+        RankedTensorType::get({newKSize, newNSize}, elementType);
+    Value paddingValue =
+        rewriter.create<ConstantOp>(loc, rewriter.getZeroAttr(elementType));
+
+    auto createPadding = [&](ArrayRef<int64_t> padding) {
```

Comment:
Partial super Nit. I would just do this

```
auto createPadding = [&](ArrayRef<int64_t> padding) {
  return llvm::to_vector<4>(llvm::map_range(padding,
      [&](int64_t dim) { return rewriter.getI64IntegerAttr(padding);}));
}
```

No reason to create a `Value` when `OpFoldResult` can just store the static value as an `Attribute`.